lib/Dialect/SPIRV/Serialization/Serializer.cpp

```diff
@@ -517,8 +547,80 @@ Serializer::processMemberDecoration(uint32_t structID, uint32_t memberIndex,
   return encodeInstructionInto(decorations, spirv::Opcode::OpMemberDecorate,
                                args);
 }
+
+LogicalResult Serializer::updateForwardCalls() {
+  for (const auto &forwardCall : forwardCalls) {
+    auto funcID = findFunctionID(std::get<0>(forwardCall));
+    if (!funcID) {
+      return emitError(std::get<2>(forwardCall),
+                       "cannot find an <id> for the function with name ")
+             << std::get<0>(forwardCall)
+             << " to update OpFunctionCallOp forward call";
+    }
+    // OpFunctionCall
+    //  word   |      word       |    word     |     word      |   word * N
+    // size|op | Resut Type <id> | Result <id> | Function <id> | arg0 <id> ... N
+    if (failed(updateInstructionWordInto(functions, std::get<1>(forwardCall), 3,
+                                         funcID))) {
+      return emitError(
+                 std::get<2>(forwardCall),
+                 "cannot update the function <id> for OpFunctionCall, "
+                 "the start address of instruction + offset is out of bound "
+                 "for the `binary` vector. Start address of OpFunctionCall ")
+             << std::get<2>(forwardCall)
+             << ", the size in words of the `binary` vector "
+             << functions.size();
+    }
+  }
+  return success();
+}
 } // namespace
 
+LogicalResult Serializer::processFunctionCallOp(spirv::FunctionCallOp op) {
+  auto funcAttrSymbolRef = op.getAttrOfType<SymbolRefAttr>("callee");
+  uint32_t resTypeID = 0;
+
+  // The number of results is checked under verifier.
+  llvm::SmallVector<Type, 1> resultTypes(op.getResultTypes());
+  if (failed(processType(op.getLoc(),
+                         (resultTypes.empty() ? getVoidType() : resultTypes[0]),
+                         resTypeID))) {
+    return failure();
+  }
+
+  auto funcName = funcAttrSymbolRef.getValue();
+  auto funcID = findFunctionID(funcName);
+  auto funcCallID = getNextID();
+  SmallVector<uint32_t, 3> operands{resTypeID, funcCallID, funcID};
+  uint32_t argIndex = 0;
+
+  for (auto *value : op.arguments()) {
+    auto valueID = findValueID(value);
+    if (!valueID) {
+      return op.emitOpError("cannot find a value for the "
+                            "argument with number ")
+             << argIndex;
+    }
+    operands.push_back(valueID);
+    ++argIndex;
+  }
+
+  if (resultTypes.size()) {
+    valueIDMap[op.getResult(0)] = funcCallID;
+  }
+
+  // The spec allows to invoke FunctionCallOp as a forward call.
```

Comment:
Instead of doing this, what about having a `getOrCreateFunctionID(StringRef)` function, which internally uses `getNextID` as the ID for that function if it is not seen yet? I think that will be simpler.

Reply:
oh, I see, thanks a lot!